diff --git a/BUILD.gn b/BUILD.gn
index 670224d..3f88e4e 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -49,6 +49,7 @@ config("no_asm_config") {
 
 # This has no sources on some platforms so must be a source_set.
 source_set("boringssl_asm") {
+
   visibility = [ ":*" ]  # Only targets in this file can depend on this.
 
   defines = []
@@ -76,7 +77,7 @@ source_set("boringssl_asm") {
   } else if (current_cpu == "x64") {
     if (is_mac) {
       sources += crypto_sources_mac_x86_64
-    } else if (is_linux || is_android) {
+    } else if (is_linux || is_android || is_ohos) {
       sources += crypto_sources_linux_x86_64
     } else {
       public_configs = [ ":no_asm_config" ]
@@ -84,14 +85,14 @@ source_set("boringssl_asm") {
   } else if (current_cpu == "x86") {
     if (is_mac) {
       sources += crypto_sources_mac_x86
-    } else if (is_linux || is_android) {
+    } else if (is_linux || is_android || is_ohos) {
       sources += crypto_sources_linux_x86
     } else {
       public_configs = [ ":no_asm_config" ]
     }
-  } else if (current_cpu == "arm" && (is_linux || is_android)) {
+  } else if (current_cpu == "arm" && (is_linux || is_android || is_ohos)) {
     sources += crypto_sources_linux_arm
-  } else if (current_cpu == "arm64" && (is_linux || is_android)) {
+  } else if (current_cpu == "arm64" && (is_linux || is_android || is_ohos)) {
     sources += crypto_sources_linux_aarch64
     asmflags += [ "-march=armv8-a+crypto" ]
   } else {
